Support RelationList field with StaticCatalogVocabulary and SelectWidget. (#4614)
plone/volto: docs/source/recipes/widget.md

Based on this setup, Volto will render this field with the `TokenWidget`.

> ## Relation fields

Reviewer: Section headings without narrative text should be avoided. They should either:

- have introductory general text describing their sub-sections, or
- be removed and their sub-sections demoted.

> 

Reviewer: ```suggestion

(single-relation-field-label)=

```

### Single relation field

> Relation field with named `StaticCatalogVocabulary`and `Select` widget:

Reviewer: ```suggestion
The following example is a single relation choice field with the named vocabulary of `StaticCatalogVocabulary` as a `Select` widget:
```

```python
relationchoice_field_named_staticcatalogvocabulary = RelationChoice(
    title="RelationChoice – named StaticCatalogVocabulary – Select widget",
    description="field/relation: relationchoice_field_named_staticcatalogvocabulary",
    vocabulary="relationchoice_field_named_staticcatalogvocabulary",
    required=False,
)
directives.widget(
    "relationchoice_field_named_staticcatalogvocabulary",
    frontendOptions={
        "widget": "select",
    },
)
```

> 

Reviewer: ```suggestion
```

It is recommended to define the vocabulary as a named `StaticCatalogVocabulary` with the field/relation name as its name.  

> This allowes the relations control panel to respect the defined restrictions to potential relation targets.

Reviewer: ```suggestion
This allows the {guilabel}`relations` control panel to respect the defined restrictions to potential relation targets.
```

{file}`vocabularies.py`
```python
from plone.app.vocabularies.catalog import StaticCatalogVocabulary
from zope.interface import provider
from zope.schema.interfaces import IVocabularyFactory

@provider(IVocabularyFactory)
def ExamplesVocabularyFactory(context=None):
    return StaticCatalogVocabulary(
        {
            "portal_type": ["example"],
            "review_state": "published",
            "sort_on": "sortable_title",
        }
    )
```

{file}`configure.zcml`
```xml
<utility
  name="relationchoice_field_named_staticcatalogvocabulary"
  component="example.contenttype.vocabularies.ExamplesVocabularyFactory"
  />
```

> 

Reviewer: ```suggestion
```

> The directive is by now default and can be ommitted.

Reviewer: ```suggestion
By following this guidance, the directive is set by default and can be omitted as shown below.
```

```python
relationchoice_field_named_staticcatalogvocabulary = RelationChoice(
    title="RelationChoice – named StaticCatalogVocabulary – Select widget",
    description="field/relation: relationchoice_field_named_staticcatalogvocabulary",
    vocabulary="relationchoice_field_named_staticcatalogvocabulary",
    required=False,
)
```

> 

Reviewer: ```suggestion

(multi-relation-field-label)=

```

### Multi relation field

> Multi relation field with named `StaticCatalogVocabulary`and `Select` widget:

Reviewer: ```suggestion
Similar to {ref}`single-relation-field-label`, the following example is a multi relation choice field with the named vocabulary of `StaticCatalogVocabulary` as a `Select` widget:
```

Based on this setup, Volto will render this field with the `TokenWidget`.

(widget-relation-field-label)=

## Relation fields

> A relation field is either a single relation field to hold at most one content object, `RelationChoice`, or a multi relation field `RelationList`, that can hold more than one content object.

Reviewer: ```suggestion
A relation field is either a single relation field to hold at most one content object, `RelationChoice`, or a multi relation field, `RelationList`, that can hold more than one content object.
```

Relation fields can be edited and rendered with the `Select` widget.

> The restriction on content types, workflow states, e.a. can be done with a `StaticCatalogVocabulary`.

Reviewer: ```suggestion
The restriction on content types, workflow states, and so on can be done with a `StaticCatalogVocabulary`.
```

> There are other vocabulary types and other widgets like the `ObjectBrowser` widget.

Reviewer: ```suggestion
There are other vocabulary types and other widgets, including the `ObjectBrowser` widget.
```

    required=False,
)
directives.widget(
    "relationchoice_field_named_staticcatalogvocabulary",
    frontendOptions={
        "widget": "select",
    },
)
```

It is recommended to define the vocabulary as a named `StaticCatalogVocabulary` with the field/relation name as its name.  
This allows the {guilabel}`relations` control panel to respect the defined restrictions to potential relation targets.

{file}`vocabularies.py`
```python
from plone.app.vocabularies.catalog import StaticCatalogVocabulary
from zope.interface import provider
from zope.schema.interfaces import IVocabularyFactory

@provider(IVocabularyFactory)
def ExamplesVocabularyFactory(context=None):
    return StaticCatalogVocabulary(
        {
            "portal_type": ["example"],
            "review_state": "published",
            "sort_on": "sortable_title",
        }
    )
```

{file}`configure.zcml`
```xml
<utility
  name="relationchoice_field_named_staticcatalogvocabulary"
  component="example.contenttype.vocabularies.ExamplesVocabularyFactory"
  />
```

The `Select` widget is currently the default for `RelationChoice` fields with vocabulary.

> Therefore the directive can be ommitted.

Reviewer: ```suggestion
Therefore the directive can be omitted.
```

> Therefore the directive can be ommitted.

Reviewer: ```suggestion
Therefore the directive can be omitted.
```
